ui-v2/app/styles/components/buttons.scss

```diff
@@ -0,0 +1,16 @@
+@import './buttons/index';
+button[type='submit'],
+a.type-create {
+  @extend %primary-button;
+}
+button[type='reset'],
+button[type='button']:not(.copy-btn):not(.type-delete),
+html.template-error div > a {
+  @extend %secondary-button;
+}
```

Comment:
Yep. Different strokes for different folks.

For me this is ideal and means my HTML is decoupled from my CSS as it should be. I don't have things like:

```html
<div class="is-small no-text-transform">
```
 which in my eyes is not far off
```html
<div style="font-size: small; text-transform: none;">
```

I'd say class names should be an extension of HTML semantics, i.e. this is what this information is, not what it looks like.

Once web components can be used, classnames will be used a lot less as you'll be able to make your own semantic elements like `<action-group>`. Each with their own shadow DOM that have their own 'imposed structure' that is editable by the component author (along with its basic styling). The user will only be able to edit styling that has been defined as editable by the author like for example its 'skin', or its content via slots (again defined by the author). This approach might help to explain a lot of your questions above (you also know how much I appreciate having the ember-block-slots addon)

A fairly good example that we've already spoken about above is an `action-group` which is is a group of 'actions' you can choose from. It happens to look like a dropdown menu in consul UI, but it could just aswell be a simple list of links (without the dropdown) but all that is controlled via CSS not the HTML.

Really `%action-group` above should be implemented like:

```html
<!-- what the information is -->
<div class="action-group"></div>

<!-- or in a web component world, or almost an ember angle brackets world -->
<action-group></action-group>
```
```css
.action-group {
/* what it looks like */
  extend: %expandable-menu;
}
```

..which is probably where I'll be going with this next.

A disadvantage of not doing this is, say for example I'd used bulma to pepper my HTML with bulma classes. Now I'm pretty much locked in to bulma and means it's probably going to take quite a bit of work to move away from bulma if I want to.

This is why it only took me a few hours max to ween myself off bulma entirely, I'd avoided the lock-in.

Reply:
To say that modifier classes aren't far off from inline styles is to say that variables aren't far off from passing in those values everywhere. They are still referential "pointers" that separate styles from markup. 

When you write deep selectors you may be separating html from css, but you are tightly coupling structure and style. In a world of web components and shadow-dom, this isn't so bad, since the shadow-doms have an isolated scope and bare markup is written less often. 

As for updating markup that is peppered with classes you no longer want, there are plenty of find/replace tools. The diff may be larger, but I don't think the amount of work is necessarily larger.